test/e2e/case18_compliance_api_test.go

```diff
@@ -112,4 +110,757 @@ var _ = Describe("Test policy webhook", Label("compliance-events-api"), Ordered,
+						severity *string
+					)
+
+					err := rows.Scan(&id, &kind, &apiGroup, &name, &ns, &pid, &spec, &specHash, &severity)
+					Expect(err).ToNot(HaveOccurred())
+
+					Expect(id).NotTo(Equal(0))
+					Expect(kind).To(Equal("ConfigurationPolicy"))
+					Expect(apiGroup).To(Equal("policy.open-cluster-management.io"))
+					Expect(ns).ToNot(BeNil())
+					Expect(*ns).To(Equal("local-cluster"))
+					Expect(pid).ToNot(BeNil())
+					Expect(*pid).ToNot(Equal(0))
+					Expect(spec).ToNot(BeNil())
+					Expect(*spec).To(Equal(`{"test":"one","severity":"low"}`))
+					Expect(specHash).ToNot(BeNil())
+					Expect(*specHash).To(Equal("cb84fe29e44202e3aeb46d39ba46993f60cdc6af"))
+					Expect(severity).ToNot(BeNil())
+					Expect(*severity).To(Equal("low"))
+
+					count++
+				}
+
+				Expect(count).To(Equal(1))
+			})
+
+			It("Should have created the event in a table", func() {
+				rows, err := db.Query("SELECT * FROM compliance_events WHERE timestamp = $1",
+					"2023-01-01T01:01:01.111Z")
+				Expect(err).ToNot(HaveOccurred())
+
+				count := 0
+				for rows.Next() {
+					var (
+						id         int
+						clusterID  int
+						policyID   int
+						compliance string
+						message    string
+						timestamp  string
+						metadata   complianceeventsapi.JSONMap
+						reportedBy *string
+					)
+
+					err := rows.Scan(&id, &clusterID, &policyID, &compliance, &message, &timestamp,
+						&metadata, &reportedBy)
+					Expect(err).ToNot(HaveOccurred())
+
+					Expect(id).NotTo(Equal(0))
+					Expect(clusterID).NotTo(Equal(0))
+					Expect(policyID).NotTo(Equal(0))
+					Expect(compliance).To(Equal("NonCompliant"))
+					Expect(message).To(Equal("configmaps [etcd] not found in namespace default"))
+					Expect(timestamp).To(Equal("2023-01-01T01:01:01.111Z"))
+					Expect(metadata).To(HaveKeyWithValue("test", true))
+					// Expect(*metadata).To(Equal(`{"test":true}`))
+					Expect(reportedBy).ToNot(BeNil())
+					Expect(*reportedBy).To(Equal("optional-test"))
+					count++
+				}
+
+				Expect(count).To(Equal(1))
+			})
+		})
+
+		Describe("POST two minimally-valid events on different clusters and policies", func() {
+			payload1 := []byte(`{
+				"cluster": {
+					"name": "cluster2",
+					"cluster_id": "test2-cluster2-fake-uuid-2"
+				},
+				"policy": {
+					"apiGroup": "policy.open-cluster-management.io",
+					"kind": "ConfigurationPolicy",
+					"name": "etcd-encryption2",
+					"spec": "{\"test\":\"two\"}"
+				},
+				"event": {
+					"compliance": "NonCompliant",
+					"message": "configmaps [etcd] not found in namespace default",
+					"timestamp": "2023-02-02T02:02:02.222Z"
+				}
+			}`)
+
+			payload2 := []byte(`{
+				"cluster": {
+					"name": "cluster3",
+					"cluster_id": "test2-cluster3-fake-uuid-3"
+				},
+				"policy": {
+					"apiGroup": "policy.open-cluster-management.io",
+					"kind": "ConfigurationPolicy",
+					"name": "etcd-encryption2",
+					"specHash": "cb84fe29e44202e3aeb46d39ba46993f60cdc6af"
```

Comment:
I think this shouldn't be allowed. Either the specHash is known in the database and if not, you must provide the spec.

Reply:
I see what you mean... we don't want an entry in the db that only has the hash, and doesn't have the full spec.

On the other hand, it feels a little weird that the user needs to know what was already in the database, or that a request might fail at one point, but then succeed later with no other changes from them (because in the meantime, something else could have put that policy spec in the database).

Reply:
I think the normal use-case is to provide `spec` and not `specHash`, however, to save on bandwidth, you can cache the `specHash` to reuse for future API usage.

In our use-case, the propagator will likely prepopulate the `parent_policies` and `policies` tables so to save on bandwidth in edge scenarios so the managed cluster just needs to send `specHash`.

Alternatively, we could have the propagator set the database ID for the `policies` table entry on the `policy-templates` entries and use that from the managed cluster side.